File: src/Python/Linked-List/86.分隔链表.py
#
# @lc app=leetcode.cn id=86 lang=python
#
# [86] 分隔链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        small_list = ListNode(0)
        small = small_list
        large_list = ListNode(0)
        large = large_list
        while head:
            if head.val < x:
                small_list.next = head
                small_list = small_list.next
            else:
                large_list.next = head
                large_list = large_list.next
            head = head.next
        large_list.next = None
        small_list.next = large.next
        return small.next
# @lc code=end

File: src/Java/Linked-List/86.分隔链表.java
/*
 * @lc app=leetcode.cn id=86 lang=java
 *
 * [86] 分隔链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode largeList = new ListNode(0);
        ListNode largeHead = largeList;
        ListNode smallList = new ListNode(0);
        ListNode smallHead = smallList;
        while (head != null) {
            if (head.val < x) {
                smallList.next = head;
                smallList = smallList.next;
            } else {
                largeList.next = head;
                largeList = largeList.next;
            }
            head = head.next;
        }
        largeList.next = null;
        smallList.next = largeHead.next;
        return smallHead.next;
    }
}
// @lc code=end

